### PR TITLE
Initial Hyena ND integration

### PR DESCRIPTION
**Pull request by dwromero-nv**
Ported initial base implementation of the HyenaND modules.

The format of data is expected to be [B, * spatial_dims, H]. 

There are tests to verify the correctness of fftconvs vs. torch.nn.ConvND. I also added a training loop for MNIST classification, which is solved (>99% acc) in 3 epochs or so. 

The most important part is how kernels are created. 
* The code creates a grid that is 2x - 1 the size of the input or input_size + 1. This is controlled by the grid_type, which can be either double or single. 
* The grid_step is used to extend / compress the grid_size if the input changes.
* The grid is centered around zero, e.g., [-1, ..., 0, ... 1] and is instantiated in multiple dimensions. 
* The grid is then passed through the kernel generator network, which is then modulated by the mask. 

**IMPORTANT**
* Causality is not properly supported yet, as the multi-dim implementation requires quite some changes relative to how Evo2 is implemented, e.g., how grids are constructed.
* No parallelization strategies are included yet.

### CODE REVIEW

**Review by farhadrgh — COMMENTED** [review]
Do you mind updating the [pre-commit pre-push config](https://github.com/NVIDIA-Digital-Bio/nvSubquadratic-private/blob/7ba8067b336b82538a2fa8dc19576466f176168a/.pre-commit-config.yaml#L39) to include the `nvsubquadratic/` directory as well for running pytest:

File: nvsubquadratic/src/hyena_nd.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+        self.gate_nonlinear = instantiate(gate_nonlinear_cfg)
+
+    def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor) -> torch.Tensor:
+        # Reshape query, key, and value to [B, C, * spatial_dims]
+        query = rearrange(query, "b ... c -> b c ...")
+        key = rearrange(key, "b ... c -> b c ...")
+        value = rearrange(value, "b ... c -> b c ...")
+
+        if not isinstance(self.short_conv, torch.nn.Identity):
```
Comment (farhadrgh) [inline nvsubquadratic/src/hyena_nd.py]: This is never False given the assert in the init, right?

File: nvsubquadratic/src/kernels_nd.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        # Initialize linear projection to be normal with mean 0 and std 2 * pi * omega_0.
+        self.linear.weight.data.normal_(mean=0.0, std=2 * torch.pi * self.omega_0)
+        if self.linear.bias is not None:
+            torch.nn.init.constant_(self.linear.bias, 0.0)
+
+        # Construct grid cache (cube) of size 2 * L_cache - 1.
+        # TODO(@dwromero): We must make sure that the grid_cache is kept in float32.
+        t = torch.linspace(-1, 1, 2 * self.L_cache - 1, dtype=torch.float32)
+        grid_cache = rearrange(torch.stack(torch.meshgrid(*[t] * data_dim, indexing="ij"), dim=-1), "... -> 1 ...")
+        self.register_buffer("grid_cache", grid_cache)
```
Comment (farhadrgh) [inline nvsubquadratic/src/kernels_nd.py]: Would we want to register buffer for 1D case?

**Review by farhadrgh — APPROVED** [review]
